Added enroll and planned graduation functionality

ntan9 opened the pull request:
Users can specify what year they enrolled HMC and what semester they plan on graduating in. Based on what year they enrolled at, the user will be able to specify (within 4 years from the enroll year) which semester they wish to graduate in, and the average remaining credits are adjusted accordingly.

Refactors all the code using `grad_year` to determine graduation requirements to now use `enroll`.

JeremyTsaii commented:
Looks pretty good! One thing is that if we aren't using gradYear anymore, we should delete that everywhere.

Also, I was testing some configurations and it seems we can get negative and infinity for remaining average credits. We should probably  have a minimum bound for the remaining average credits to 0. Not sure how infinity came up, perhaps a divide by 0?

JeremyTsaii commented on src/components/InfoModal.tsx:
```diff
@@ -89,19 +89,52 @@ function InfoModal(): JSX.Element {
   const nameRef = useRef('')
 
-  const [school, setSchool] = useState(schoolName)
+  const [school, setSchool] = useState(schoolDict[schoolName])
```
I believe the issue was fixed in the last push to master. Let's use what's currently in the master branch.

JeremyTsaii commented on src/components/InfoModal.tsx:
```diff
@@ -89,19 +89,52 @@ function InfoModal(): JSX.Element {
+
+  const updateGradDates = (start: number): { value: string }[] => {
+    return [
+      {
```
Nice, I like that it's capped at 4 years. In the future, we may consider more but this is good for now.

ntan9 commented:
Yes good point on that. I think I removed `grad_years` everywhere except for the Graphql queries, so I can fix that.

As for the infinity on the average remaining credits, that happens regardless when you still need credits to graduate by the time you intend on graduating. For example, if you enroll in 2018 and your planned semester is Spring of 2022, you need to graduate within 8 semesters. If you fill in all the year cards but still have credits left over, it'll indicate you have an infinite number of credits left. I honestly think that makes sense, as it indicates that it's impossible for you to graduate within the desired number of semesters.

As for negative, this happens when you have more semesters filled out than the number of semesters that you intend on doing (indicated through enroll and planned graduation). It is possible to instead make it result to infinity, which will indicate, as above, that it's impossible for the user to graduate within the desired window of time.

JeremyTsaii commented:
Gotcha, I see. Okay, I think the infinity thing is fine, but we should make negatives to infinities to indicate that it's not possible to graduate with the specified enroll/grad selection. We should store infinity inside the db, but we should also indicate to the user that infinity means impossible. Could you add something like "Infinity - not possible" or something along those lines to the UI?

ntan9 commented:
Good idea. I've got the average remaining credits stats to report "Infinity - Not Possible" if they can't graduate within the planned time, "Graduating after `plannedGrad`" if they have the necessary credits but not within the desired number of semesters, "0" if they don't need anymore credits, and the average otherwise.

Also, I added a fix to the remaining credits stats, so that it now has a minimum value of 0. Before, it was able to indicate that a user had a negative number of credits left.

Lastly, I added a fix to the major requirements for the CS majors. I added one more semester of CS 195, which is part of the major requirements (in the non-pandemic times).

JeremyTsaii commented on src/components/RightStatsCard.tsx:
```diff
@@ -174,12 +179,39 @@ const calculateStats = (school: string, courses: Courses[]): Stats => {
-  statsObj.rem = requiredCredits - totalCredits
+
+  // Set minimum remaining credits to 0
+  if (totalCredits > requiredCredits) {
```
Change to ternary for conciseness/consistence.

JeremyTsaii commented on src/components/RightStatsCard.tsx:
```diff
@@ -174,12 +179,39 @@ const calculateStats = (school: string, courses: Courses[]): Stats => {
+  if (totalSemesters <= semesters && statsObj.rem > 0) {
+    // If user still has remaining credits within time to graduate
+    statsObj.avgRem = 'Infinity - Not Possible'
+  } else if (totalSemesters < semesters && statsObj.rem <= 0) {
```
At this point, `statsObj.rem` should be greater than or equal to 0 right?

ntan9 replied:
Yes, that would be more concise.
